RUFAS/routines/animal/life_cycle/heiferII.py

```diff
@@ -310,390 +337,983 @@ def update(self, sim_day):
+        ----------
+        sim_day : int
+            The current day of the entire simulation.
+
+        Returns
+        -------
+        None
+        """
+
+        self.log_event(self.days_born, sim_day, const.ESTRUS_OCCURRED_NOTE)
+        is_estrus_detected = self._detect_estrus(self._get_user_defined_or_default_synch_ed_estrus_detection_rate())
+        if is_estrus_detected:
+            self.log_event(self.days_born, sim_day, const.ESTRUS_DETECTED_NOTE)
+            self.conception_rate = self._get_user_defined_TAI_conception_rate()
+            self.ai_day = self.days_born + 1
+            self.log_event(self.days_born, sim_day, f'{const.AI_DAY_SCHEDULED_NOTE} on day {self.ai_day}')
+        else:
+            self._handle_estrus_not_detected_in_synch_ed(sim_day)
+
+    def _handle_estrus_not_detected_in_synch_ed(self, sim_day: int) -> None:
+        """
+        Handle the scenario where estrus is not detected in the heifers in the SynchED program.
+
+        Parameters
+        ----------
+        sim_day : int
+            The current day of the entire simulation.
+
+        Returns
+        -------
+        None
+        """
+
+        self.log_event(self.days_born, sim_day, const.ESTRUS_NOT_DETECTED_NOTE)
+        self.log_event(self.days_born, sim_day, const.TAI_AFTER_ESTRUS_NOT_DETECTED_IN_SYNCH_ED_NOTE)
+        internal_fallback_protocol = InternalReproSettings.HEIFER_REPRO_PROTOCOLS[
+            self._get_user_defined_or_default_repro_sub_protocol()]['when_estrus_not_detected']
+
+        self._set_repro_program(sim_day, internal_fallback_protocol['repro_protocol'])
+        self._set_up_hormone_schedule('heifers', internal_fallback_protocol['repro_sub_protocol'], self.days_born)
+        self._TAI_conception_rate = internal_fallback_protocol['repro_sub_properties']['conception_rate']
+        self._execute_hormone_delivery_schedule(sim_day, self._hormone_schedule)
+
+    def _set_repro_program(self, sim_day: int, repro_program: Literal['ED', 'TAI', 'SynchED']) -> None:
+        """
+        Set the reproduction program for the heifer.
+
+        Parameters
+        ----------
+        sim_day : int
+            The current day of the entire simulation.
+        repro_program : Literal['ED', 'TAI', 'SynchED']
+            The reproduction program to set.
+
+        Returns
+        -------
+        None
+        """
+
+        if repro_program not in [HeiferReproProtocolEnum.ED.value, HeiferReproProtocolEnum.TAI.value,
+                                 HeiferReproProtocolEnum.SynchED.value]:
+            raise ValueError(f'Invalid repro program: {repro_program}')
+
+        if self.repro_program == repro_program:
+            return
+
+        self.log_event(self.days_born, sim_day,
+                       f'{const.SETTING_REPRO_PROGRAM_NOTE} from {self.repro_program} '
+                       f'to {repro_program}')
+        self.repro_program = repro_program
+
+    def open(self, sim_day: int) -> None:
```

Comment:
```suggestion
    def open_heiferII(self, sim_day: int) -> None:
```
I think this makes it clearer what it is opening.

Reply:
I will hold off on this one, since in the cow repro refactor, there is also another `open` method. There is a potential for method overriding here.